File: Deep-Learning-for-Beginners-master/Matlab/ch.6/PlotFeatures.m
% 该文件展示了在正向传播的过程中图像是如何一步步被处理的。

clear all

load('MnistConv.mat') % 加载预训练模型

k  = 2; % 取训练集中第 2 张图片作为例子进行说明
x  = X(:, :, k);
y1 = Conv(x, W1);
y2 = ReLU(y1);
y3 = Pool(y2);
y4 = reshape(y3, [], 1);
v5 = W5*y4;
y5 = ReLU(v5);
v  = Wo*y5;
y  = Softmax(v);


figure;
display_network(x(:)); % 显示输入的图片
title('Input Image')

convFilters = zeros(9*9, 20);
for i = 1:20
  filter            = W1(:, :, i);
  convFilters(:, i) = filter(:);
end
figure
display_network(convFilters); % 显示卷积层用到的过滤器
title('Convolution Filters')

fList = zeros(20*20, 20);
for i = 1:20
  feature     = y1(:, :, i);
  fList(:, i) = feature(:); % 显示图片经过卷积处理后的结果
end
figure
display_network(fList);
title('Features [Convolution]')

fList = zeros(20*20, 20);
for i = 1:20
  feature     = y2(:, :, i);
  fList(:, i) = feature(:);
end
figure
display_network(fList); % 显示图片经过卷积 + ReLU 处理后的结果
title('Features [Convolution + ReLU]')

fList = zeros(10*10, 20);
for i = 1:20
  feature     = y3(:, :, i);
  fList(:, i) = feature(:);
end
figure
display_network(fList); % 显示图片经过卷积 + ReLU + 池化处理后的结果
title('Features [Convolution + ReLU + MeanPool]')

File: Deep-Learning-for-Beginners-master/R/ch.6/PlotFeatures.R
rm(list=ls())

library(plot.matrix)
source("./Conv.R")
source("./ReLU.R")
source("./Pool.R")
source("./Softmax.R")

load("./.RData")

k <- 2
x <- X[, , k]                               # Input,       28x28
y1 <- Conv(x, W1)                           # Convolution, 20x20x20
y2 <- ReLU(y1)                              #
y3 <- Pool(y2)                              # Pool,        10x10x20
y4 <- as.vector(y3)                         #              2000
v5 <- W5 %*% y4                             # ReLU,        100
y5 <- ReLU(v5)                              #
v <- Wo %*% y5                              # Softmax,     10
y <- Softmax(v)                             #

plot(x)

convFilters <- rbind(
  cbind(W1[, , 1], W1[, , 2], W1[, , 3], W1[, , 4], W1[, , 5]),
  cbind(W1[, , 6], W1[, , 7], W1[, , 8], W1[, , 9], W1[, , 10]),
  cbind(W1[, , 11], W1[, , 12], W1[, , 13], W1[, , 14], W1[, , 15]),
  cbind(W1[, , 16], W1[, , 17], W1[, , 18], W1[, , 19], W1[, , 20])
)
plot(convFilters)

fList <- rbind(
  cbind(y1[, , 1], y1[, , 2], y1[, , 3], y1[, , 4], y1[, , 5]),
  cbind(y1[, , 6], y1[, , 7], y1[, , 8], y1[, , 9], y1[, , 10]),
  cbind(y1[, , 11], y1[, , 12], y1[, , 13], y1[, , 14], y1[, , 15]),
  cbind(y1[, , 16], y1[, , 17], y1[, , 18], y1[, , 19], y1[, , 20])
)
plot(fList)

fList <- rbind(
  cbind(y2[, , 1], y2[, , 2], y2[, , 3], y2[, , 4], y2[, , 5]),
  cbind(y2[, , 6], y2[, , 7], y2[, , 8], y2[, , 9], y2[, , 10]),
  cbind(y2[, , 11], y2[, , 12], y2[, , 13], y2[, , 14], y2[, , 15]),
  cbind(y2[, , 16], y2[, , 17], y2[, , 18], y2[, , 19], y2[, , 20])
)
plot(fList)

fList <- rbind(
  cbind(y3[, , 1], y3[, , 2], y3[, , 3], y3[, , 4], y3[, , 5]),
  cbind(y3[, , 6], y3[, , 7], y3[, , 8], y3[, , 9], y3[, , 10]),
  cbind(y3[, , 11], y3[, , 12], y3[, , 13], y3[, , 14], y3[, , 15]),
  cbind(y3[, , 16], y3[, , 17], y3[, , 18], y3[, , 19], y3[, , 20])
)
plot(fList)
